### PR TITLE
Norm-Based Adaptive Moment Estimation with Orthogonalized Momentum

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
Build Namo as another method to normalize Muon updates 

From Adam improves Muon (https://arxiv.org/abs/2602.17080)

### CODE REVIEW

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 1]
```diff
@@ -70,7 +73,7 @@ def __init__(
         scale_mode: muon.MuonScaleT = "spectral",
         extra_scale_factor: float = 1.0,
         use_syrk: bool = False,
-        moment2_method: Literal["adamuon", "normuon"] = "adamuon",
+        moment2_method: Literal["adamuon", "normuon", "namo"] = "adamuon",
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Nit: Define a type since now it has three variant. Can also do it when we have more.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 2]
```diff
@@ -121,6 +125,9 @@ def _initialize_moment2(
                 moment2_shape = list(grad.shape)
                 moment2_shape[avg_dim] = 1
                 moment2 = torch.zeros(moment2_shape, dtype=grad.dtype, device=grad.device)
+            elif self.moment2_method == "namo":
+                # Scalar second moment: EMA of ||G_t||_F^2
+                moment2 = torch.zeros((), dtype=grad.dtype, device=grad.device)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Q: Why empty tensor? seems a meaningless placeholder.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 3]
```diff
@@ -132,22 +139,37 @@ def _apply_moment2_normalization(
         moment2: torch.Tensor,
         beta2: float,
         eps: float,
+        *,
+        grad_fro_sq: torch.Tensor | None = None,
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: I think we always default to P2 norm for flattened. `grad_norm_square` would be a better name given grad_norm is so widely used.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 4]
```diff
@@ -132,22 +139,37 @@ def _apply_moment2_normalization(
...
+        pre_orth_norm: torch.Tensor | None = None,
     ) -> torch.Tensor:
-        """Apply AdamW-style second moment accumulation and normalization.
+        """Apply second moment accumulation and normalization.
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py:164]: Good catch.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 5]
```diff
@@ -132,22 +139,37 @@ def _apply_moment2_normalization(
         moment2: torch.Tensor,
         beta2: float,
         eps: float,
+        *,
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py:160]: Comment: Flag based arguments are never great. Fine to user internally to some extend. But if there are more flag dependent arguments are needed, should redesign the interface.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 6]
```diff
@@ -175,6 +197,17 @@ def _apply_moment2_normalization(
             step_size = moment2.clamp_min(eps).rsqrt_()
             return orth_grad * step_size
 
+        elif self.moment2_method == "namo":
+            assert grad_fro_sq is not None and pre_orth_norm is not None, "NAMO requires grad_fro_sq and pre_orth_norm"
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Fix: Raise RuntimeError here, don't assert.
Check style guide (google) for rules of assert.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 7]
```diff
@@ -231,15 +269,26 @@ def step(self, closure: Callable[[], float] | None = None) -> float | None:
                     group_kwargs = {k: v for k, v in group.items() if k != "params"}
                     orth_grad = self.orthogonalize(p, grad, **group_kwargs)
 
-                update = self._apply_moment2_normalization(
-                    orth_grad=orth_grad,
-                    moment2=state["moment2_buffer"],
-                    beta2=group["beta2"],
-                    eps=group["eps"],
-                )
+                if self.moment2_method == "namo":
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Q: why this dispatch is outside _apply_moment2_normalization? should be handled inside it if no particular reason.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 8]
```diff
@@ -231,15 +269,26 @@ def step(self, closure: Callable[[], float] | None = None) -> float | None:
...
+                    # Capture ||pre_orth_grad||_F (after momentum + Nesterov)
+                    update = self._apply_moment2_normalization(
+                        orth_grad=orth_grad,
+                        moment2=state["moment2_buffer"],
+                        beta2=group["beta2"],
+                        eps=group["eps"],
+                        grad_fro_sq=grad_fro_sq,
+                        pre_orth_norm=torch.linalg.vector_norm(grad),
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Q: grad_fro_sq and pre_orth_norm seems to be the same here?

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 9]
```diff
@@ -231,15 +269,26 @@ def step(self, closure: Callable[[], float] | None = None) -> float | None:
...
+                    update = self._apply_moment2_normalization(
+                        orth_grad=orth_grad,
+                        moment2=state["moment2_buffer"],
+                        beta2=group["beta2"],
+                        eps=group["eps"],
+                    )
 
                 # perform weight update
-                # scale is applied to have update RMS == 1
+                # scale is applied to have update RMS roughly 1
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: It is not clear what scale means in this line. I'd just remove the comment including the line above.

File: tests/test_adaptive_muon.py [hunk 10]
```diff
@@ -93,6 +94,9 @@ def test_second_moment_matches_shapes(self, shape, second_moment_method) -> None
             expected_shape = list(shape)
             expected_shape[avg_dim] = 1
             self.assertEqual(list(second_moment.shape), expected_shape)
+        elif second_moment_method == "namo":
+            # Scalar buffer
```
Comment (skyw) [inline tests/test_adaptive_muon.py]: Fix: meaningless and confusing comment. Remove or explain more what it asserts.

**Review by skyw — CHANGES_REQUESTED** [review]

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 1 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Moment2MethodT type alias (L10-12): Extracted the Literal["adamuon", "normuon", "namo"] into a named type and exported it in __all__.

Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Class docstring updated (L17-27): Now mentions NAMO alongside AdaMuon/NorMuon, and removes the "AdamW-style or NorMuon-style" phrasing.

Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Signature uses type alias (L36): moment2_method: Moment2MethodT instead of inline Literal.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 2 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: NAMO scalar buffer (L44-46): torch.tensor(0.0, ...) instead of torch.zeros(()) for clarity.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 3 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Renamed parameters (L54-57, L65-71): grad_fro_sq / pre_orth_norm became raw_grad / pre_orth_grad -- pass the raw tensors, not precomputed norms.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 5 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py:160]: assert replaced with raise RuntimeError (L79-81): Per Google style guide.

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 7 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Norm computation moved inside _apply_moment2_normalization (L85-90): vector_norm(...).square() and vector_norm(...) are now internal to the method rather than computed at the call site.

Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Unified call site in step() (L99-138): Removed the if self.moment2_method == "namo" dispatch. Now there's a single _apply_moment2_normalization(...) call with raw_grad and pre_orth_grad passed through (they're None for non-NAMO methods).

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 9 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: Removed unclear comment (L129-130): "perform weight update / scale is applied to have update RMS roughly 1".

File: tests/test_adaptive_muon.py (hunk 10 above)
Comment (mkhona-nvidia) [reply tests/test_adaptive_muon.py]: done

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 11]
```diff
@@ -121,6 +127,8 @@ def _initialize_moment2(
                 moment2_shape = list(grad.shape)
                 moment2_shape[avg_dim] = 1
                 moment2 = torch.zeros(moment2_shape, dtype=grad.dtype, device=grad.device)
+            elif self.moment2_method == "namo":
+                moment2 = torch.tensor(0.0, dtype=grad.dtype, device=grad.device)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: In general, avoid empty shape tensors. see example below
```python
In [2]: a = torch.zeros(1)
In [3]: b = torch.tensor(0)

In [4]: b
Out[4]: tensor(0)
In [5]: a
Out[5]: tensor([0.])

In [6]: a.shape
Out[6]: torch.Size([1])
In [7]: b.shape
Out[7]: torch.Size([])
```
It is less error prone to compare size 1 instead of empty shape.

**Review by skyw — DISMISSED** [review]
Recommend to remove empty tensor shape usage. otherwise LGTM

**Review by skyw — DISMISSED** [review]

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py [hunk 12]
```diff
@@ -132,16 +138,16 @@ def _init_group(
...
                         raise ValueError(
                             f"{self.__class__.__name__} only supports 2D parameters, got shape {tuple(p.data.shape)}"
                         )
                     avg_dim = -1 if p.data.shape[-2] >= p.data.shape[-1] else -2
-                    # Specify the shape with reduced dimension
                     moment2_shape = list(p.data.shape)
                     moment2_shape[avg_dim] = 1
                     state["moment2_buffer"] = torch.zeros(moment2_shape, dtype=p.data.dtype, device=p.data.device)
+                elif self.moment2_method == "namo":
+                    state["moment2_buffer"] = torch.tensor(0.0, dtype=p.data.dtype, device=p.data.device)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: What was he reason to rollback?

**Review by skyw — CHANGES_REQUESTED** [review]

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 12 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: it isn't a rollback, it was an empty tensor and now it is a scalar initialized to 0

Comment (skyw) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: There was a version changed to use zeros per previous suggestion. test needs to be updated accordingly

**Review by skyw — CHANGES_REQUESTED** [review]

File: emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py (hunk 12 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py]: good point, let me revert

**Review by skyw — APPROVED** [review]
